### PR TITLE
Improve spacy.gold (no GoldParse, no json format!)

### PR DESCRIPTION
**Pull request by honnibal**
~I'm doing some exploratory work on improving the `GoldCorpus` stuff. The first thing I tried was to break it up into a subpackage to make it a bit easier to work with. I'm still exploring other changes but if we want we could merge this first. It shouldn't affect anything, but I think it will decrease compile times a bit.~

Edit: Okay this turned into bigger changes! But very nice...

* Use the `Doc` object for annotations. This means the `Example` class just holds two `Doc` objects and an alignment.
* Remove the `GoldParse` class. Equivalent functionality is now just internals inside the parser transition systems.
* Train from the `DocBin` binary format. This is smaller, faster, and removes the need for the JSON format.
* Update the converters to be able to output `DocBin`.
* Simplify the `GoldCorpus` class and rename to just `Corpus`.
* Improve performance of the `doc.from_array` method, to further increase data reading speed.
* Update and improve efficiency of the parser oracles

There are still some things left to do.

* [x] Fix segfault in parser training when `gold_preproc`
* [x] Restore the `gold_preproc` functionality, if we want that.
* [x] Finish updating the NER oracle
* [x] Finish updating tests
* [x] Double-check NER alignment and training
* [ ] Restore the data augmentation functionality

## Checklist

- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by adrianeboyd** [comment]
I really don't think `GoldParse` should be under `spacy.syntax`. It could stand to be renamed, but I think it really belongs under `spacy.gold`?

**Comment by honnibal** [comment]
I want to try to get it out of the API and make it just something that the parser and NER use internally. In the meantime it does import from `spacy.gold` so it can be found where it used to be.

But my idea is that if we just add let the `Example` class have the alignment, and add a method `example.get_aligned(field)` to get aligned data for a single field, we'll avoid a lot of problems. Each component could control how it wants to adjust the gold-standard for stuff like misalignments, and if you only have a text categorizer the alignment won't need to be calculated at all.

We'd also eliminate all the mess from the `GoldCorpus`. The `GoldCorpus` will just return `Example` objects, and it'll be up to the components to decide what they want to do.

**Comment by honnibal** [comment]
I suppose you're right that in the absence of other changes it should be under `spacy.gold`

**Comment by honnibal** [comment]
The idea now is to use the `Doc` class for the annotations as well, and store the alignment with it in the `Example` class. This will avoid the whole json format nastiness.

File: spacy/tokens/token.pyx [hunk 1]
```diff
@@ -787,8 +791,7 @@ cdef class Token:
 
         RETURNS (str): IOB code of named entity tag.
         """
-        iob_strings = ("", "I", "O", "B")
-        return iob_strings[self.c.ent_iob]
+        return self.iob_strings()[self.c.ent_iob]
```
Comment (honnibal) [inline spacy/tokens/token.pyx:794]: This stuff is kind of frustrating every time. Maybe we should have a `token.iob_tag` attribute?

Comment (svlandeg) [reply spacy/tokens/token.pyx:794]: I'm not sure I understand what you mean? `token.ent_iob_` does give the IOB tag. But I guess the question is, who should be responsible of making the conversion from int to string?

File: spacy/tokens/_serialize.py [hunk 2]
```diff
@@ -9,6 +9,19 @@
...
 
+ALL_ATTRS = (
+    "ORTH",
+    "TAG",
+    "HEAD",
+    "DEP",
+    "SENT_START",
+    "ENT_IOB",
+    "ENT_TYPE",
+    "LEMMA",
```
Comment (adrianeboyd) [inline spacy/tokens/_serialize.py]: Add "POS" here, too.

Comment (adrianeboyd) [reply spacy/tokens/_serialize.py]: Nevermind...

File: spacy/tokens/token.pyx (hunk 1 above)
Comment (honnibal) [reply spacy/tokens/token.pyx:794]: I just meant it's annoying to assemble the tag from the IOB component and the ent_type component. It'd be kind of nice to have an attribute that just did it.

**Pull request merged**